ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapManager.java

```diff
@@ -261,45 +283,151 @@ private boolean checkEnoughUnallocatedRawCapacity(int replicaCountPerDatacenter,
+      throw new IllegalArgumentException("retryIfNotRackAware is false, but the datacenter: " + datacenter.getName()
+          + " does not have rack information");
+    } else {
+      disks = allocateDisksForPartition(replicaCountPerDatacenter, replicaCapacityInBytes, datacenter, false);
+    }
+
+    if (disks.size() < replicaCountPerDatacenter) {
+      System.err.println(
+          "Could only allocate " + disks.size() + "/" + replicaCountPerDatacenter + " replicas in datacenter: "
+              + datacenter.getName());
     }
-    return allocatedDisks;
+    return disks;
   }
 
-  // Best effort (or less) allocation of partitions. I.e., size of returned list may be less than numPartitions.
+  /**
+   * Allocate partitions for {@code numPartitions} new partitions on all datacenters.
+   *
+   * @param numPartitions How many partitions to allocate.
+   * @param replicaCountPerDatacenter The number of replicas per partition on each datacenter
+   * @param replicaCapacityInBytes How large each replica (of a partition) should be
+   * @param retryIfNotRackAware {@code true} if we should attempt a non rack-aware allocation if a rack-aware one
+   *                            is not possible.
+   * @return A list of the new {@link PartitionId}s.
+   */
   public List<PartitionId> allocatePartitions(int numPartitions, int replicaCountPerDatacenter,
-      long replicaCapacityInBytes) {
+      long replicaCapacityInBytes, boolean retryIfNotRackAware) {
     ArrayList<PartitionId> partitions = new ArrayList<PartitionId>(numPartitions);
 
     while (checkEnoughUnallocatedRawCapacity(replicaCountPerDatacenter, replicaCapacityInBytes) && numPartitions > 0) {
-      List<Disk> disks =
-          allocateDisksForPartition(replicaCountPerDatacenter, replicaCapacityInBytes, hardwareLayout.getDatacenters());
-      if (disks.size() == 0) {
-        System.err.println("numPartitions: " + numPartitions);
-        break;
+      List<Disk> disksToAllocate = new ArrayList<>();
+      for (Datacenter datacenter : hardwareLayout.getDatacenters()) {
+        List<Disk> disks =
+            allocateDisksForPartitionWithPotentialRetry(replicaCountPerDatacenter, replicaCapacityInBytes, datacenter,
+                retryIfNotRackAware);
+        disksToAllocate.addAll(disks);
       }
-      partitions.add(partitionLayout.addNewPartition(disks, replicaCapacityInBytes));
+      partitions.add(partitionLayout.addNewPartition(disksToAllocate, replicaCapacityInBytes));
       numPartitions--;
     }
 
     return partitions;
   }
 
-  public void addReplicas(PartitionId partitionId, String dataCenterName) {
+  /**
+   * Add a set of replicas on a new datacenter for an existing partition.
+   *
+   * @param partitionId The partition to add the new datacenter to
+   * @param dataCenterName The name of the new datacenter
+   * @param retryIfNotRackAware {@code true} if we should attempt a non rack-aware allocation if a rack-aware one
```

Comment:
drop the first person subjective: "if a non rack-aware allocation should be attempted..."